pychromecast/controllers/spotify.py

```diff
@@ -38,16 +40,34 @@ def receive_message(self, _message, data: dict):
 
         Called when a message is received.
         """
-        if data["type"] == TYPE_SET_CREDENTIALS_RESPONSE:
-            self.send_message({"type": TYPE_GET_INFO, "payload": {}})
-        if data["type"] == TYPE_SET_CREDENTIALS_ERROR:
-            self.device = None
-            self.credential_error = True
-            self.waiting.set()
         if data["type"] == TYPE_GET_INFO_RESPONSE:
             self.device = data["payload"]["deviceID"]
+            self.client = data["payload"]["clientID"]
+            headers = {
+                'authority': 'spclient.wg.spotify.com',
+                'authorization': 'Bearer {}'.format(self.access_token),
+                'content-type': 'text/plain;charset=UTF-8'
+            }
+
+            request_body = json.dumps({'clientId': self.client, 'deviceId': self.device})
+
+            response = requests.post('https://spclient.wg.spotify.com/device-auth/v1/refresh', headers=headers, data=request_body)
```

Comment:
Can you test #509 which doesn't use private APIs to see if it works?

Reply:
@fondberg 

The main problem I see is that the authToken used in the addUser request is not the same authToken that is retrieved via the sp_dc and sp_key values. 

I figured out calling that api from looking at whats happening in the browser. If I just use the some authToken thats passed in I actually get a addUserError response.

Reply:
I just tried the code in #509 and I get a timeout.

```
home-assistant   | 2021-08-15 11:31:52 ERROR (MainThread) [homeassistant.core] Error executing service: <ServiceCall spotcast.start (c:ab17c97250473019a4921369e4715140): uri=spotify:playlist:37i9dQZF1DXaa8UmWJHYTU?si=dbed7b69bfdb4e6c, entity_id=media_player.bedroom_speaker, force_playback=False, offset=0, ignore_fully_played=False, shuffle=False, start_volume=101, random_song=False, repeat=off>
home-assistant   | Traceback (most recent call last):
home-assistant   |   File "/usr/src/homeassistant/homeassistant/core.py", line 1507, in catch_exceptions
home-assistant   |     await coro_or_task
home-assistant   |   File "/usr/src/homeassistant/homeassistant/core.py", line 1530, in _execute_service
home-assistant   |     await self._hass.async_add_executor_job(handler.job.target, service_call)
home-assistant   |   File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
home-assistant   |     result = self.fn(*self.args, **self.kwargs)
home-assistant   |   File "/config/custom_components/spotcast/__init__.py", line 143, in start_casting
home-assistant   |     spotify_device_id = spotcast_controller.get_spotify_device_id(
home-assistant   |   File "/config/custom_components/spotcast/spotcast_controller.py", line 202, in get_spotify_device_id
home-assistant   |     spotify_cast_device.startSpotifyController(access_token, expires)
home-assistant   |   File "/config/custom_components/spotcast/spotcast_controller.py", line 84, in startSpotifyController
home-assistant   |     sp.launch_app()
home-assistant   |   File "/usr/local/lib/python3.9/site-packages/pychromecast/controllers/spotify.py", line 87, in launch_app
home-assistant   |     raise LaunchError(
home-assistant   | pychromecast.error.LaunchError: Timeout when waiting for status response from Spotify app

```

Reply:
Trying the code in #509 

Gives me an addUserError

```
home-assistant   | 2021-08-15 11:40:14 DEBUG (Thread-7) [pychromecast.socket_client] [Bedroom speaker(192.168.0.162):8009] Received: Message urn:x-cast:com.spotify.chromecast.secure.v1 from f59a660b-975b-4f40-9c55-e61d0fa040da to sender-0: {'type': 'addUserError', 'payload': {'status': 202}}
home-assistant   | 2021-08-15 11:40:14 ERROR (MainThread) [homeassistant.components.script.j_lo_fi] J Lo-Fi: Error executing script. Unexpected error for call_service at pos 1: Timeout when waiting for status response from Spotify app
home-assistant   | Traceback (most recent call last):
home-assistant   |   File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 371, in _async_step
home-assistant   |     await getattr(self, handler)()
home-assistant   |   File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 571, in _async_call_service_step
home-assistant   |     await service_task
home-assistant   |   File "/usr/src/homeassistant/homeassistant/core.py", line 1491, in async_call
home-assistant   |     task.result()
home-assistant   |   File "/usr/src/homeassistant/homeassistant/core.py", line 1530, in _execute_service
home-assistant   |     await self._hass.async_add_executor_job(handler.job.target, service_call)
home-assistant   |   File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
home-assistant   |     result = self.fn(*self.args, **self.kwargs)
home-assistant   |   File "/config/custom_components/spotcast/__init__.py", line 143, in start_casting
home-assistant   |     spotify_device_id = spotcast_controller.get_spotify_device_id(
home-assistant   |   File "/config/custom_components/spotcast/spotcast_controller.py", line 202, in get_spotify_device_id
home-assistant   |     spotify_cast_device.startSpotifyController(access_token, expires)
home-assistant   |   File "/config/custom_components/spotcast/spotcast_controller.py", line 84, in startSpotifyController
home-assistant   |     sp.launch_app()
home-assistant   |   File "/usr/local/lib/python3.9/site-packages/pychromecast/controllers/spotify.py", line 87, in launch_app
home-assistant   |     raise LaunchError(
home-assistant   | pychromecast.error.LaunchError: Timeout when waiting for status response from Spotify app
home-assistant   | 2021-08-15 11:40:14 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [139961477371360] Timeout when waiting for status response from Spotify app
home-assistant   | Traceback (most recent call last):
home-assistant   |   File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
home-assistant   |     await hass.services.async_call(
home-assistant   |   File "/usr/src/homeassistant/homeassistant/core.py", line 1491, in async_call
home-assistant   |     task.result()
home-assistant   |   File "/usr/src/homeassistant/homeassistant/core.py", line 1526, in _execute_service
home-assistant   |     await handler.job.target(service_call)
home-assistant   |   File "/usr/src/homeassistant/homeassistant/components/script/__init__.py", line 278, in service_handler
home-assistant   |     await script_entity.async_turn_on(
home-assistant   |   File "/usr/src/homeassistant/homeassistant/components/script/__init__.py", line 382, in async_turn_on
home-assistant   |     await coro
home-assistant   |   File "/usr/src/homeassistant/homeassistant/components/script/__init__.py", line 409, in _async_run
home-assistant   |     return await self.script.async_run(script_vars, context)
home-assistant   |   File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 1219, in async_run
home-assistant   |     await asyncio.shield(run.async_run())
home-assistant   |   File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 353, in async_run
home-assistant   |     await self._async_step(log_exceptions=False)
home-assistant   |   File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 371, in _async_step
home-assistant   |     await getattr(self, handler)()
home-assistant   |   File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 571, in _async_call_service_step
home-assistant   |     await service_task
home-assistant   |   File "/usr/src/homeassistant/homeassistant/core.py", line 1491, in async_call
home-assistant   |     task.result()
home-assistant   |   File "/usr/src/homeassistant/homeassistant/core.py", line 1530, in _execute_service
home-assistant   |     await self._hass.async_add_executor_job(handler.job.target, service_call)
home-assistant   |   File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
home-assistant   |     result = self.fn(*self.args, **self.kwargs)
home-assistant   |   File "/config/custom_components/spotcast/__init__.py", line 143, in start_casting
home-assistant   |     spotify_device_id = spotcast_controller.get_spotify_device_id(
home-assistant   |   File "/config/custom_components/spotcast/spotcast_controller.py", line 202, in get_spotify_device_id
home-assistant   |     spotify_cast_device.startSpotifyController(access_token, expires)
home-assistant   |   File "/config/custom_components/spotcast/spotcast_controller.py", line 84, in startSpotifyController
home-assistant   |     sp.launch_app()
home-assistant   |   File "/usr/local/lib/python3.9/site-packages/pychromecast/controllers/spotify.py", line 87, in launch_app
home-assistant   |     raise LaunchError(
home-assistant   | pychromecast.error.LaunchError: Timeout when waiting for status response from Spotify app
```